test/serving.bash

```diff
@@ -0,0 +1,208 @@
+
+  go_test_e2e -tags=preupgrade -timeout=20m ./test/upgrade \
+    --imagetemplate "$image_template" \
+    --kubeconfig "$KUBECONFIG" \
+    --resolvabledomain || return 1
+
+  logger.info "Starting prober test"
+
+  rm -f /tmp/prober-signal
+  go_test_e2e -tags=probe -timeout=20m ./test/upgrade \
+    --imagetemplate "$image_template" \
+    --kubeconfig "$KUBECONFIG" \
+    --resolvabledomain &
+
+  # Wait for the upgrade-probe kservice to be ready before proceeding
+  timeout 900 '[[ $(oc get services.serving.knative.dev upgrade-probe -n serving-tests -o=jsonpath="{.status.conditions[?(@.type==\"Ready\")].status}") != True ]]' || return 1
+
+  PROBER_PID=$!
+
+  if [[ $UPGRADE_SERVERLESS == true ]]; then
+    serving_version=$(oc get knativeserving.operator.knative.dev knative-serving -n $SERVING_NAMESPACE -o=jsonpath="{.status.version}")
+
+    # Get latest CSV from the given channel
+    upgrade_to=$("${rootdir}/hack/catalog.sh" | sed -n '/channels/,$p;' | sed -n "/- name: \"${OLM_UPGRADE_CHANNEL}\"$/{n;p;}" | awk '{ print $2 }')
+
+    cluster_version=$(oc get clusterversion -o=jsonpath="{.items[0].status.history[?(@.state==\"Completed\")].version}")
+    if [[ "$cluster_version" = 4.1.* || "${HOSTNAME}" = *ocp-41* || \
+          "$cluster_version" = 4.2.* || "${HOSTNAME}" = *ocp-42* ]]; then
+      if approve_csv "$upgrade_to" "$OLM_UPGRADE_CHANNEL" ; then # Upgrade should fail on OCP 4.1, 4.2
+        return 1
+      fi
+      # Check we got RequirementsNotMet error
+      [[ $(oc get ClusterServiceVersion $upgrade_to -n $OPERATORS_NAMESPACE -o=jsonpath="{.status.requirementStatus[?(@.name==\"$upgrade_to\")].message}") =~ "requirement not met: minKubeVersion" ]] || return 1
+      # Check KnativeServing still has the old version
+      [[ $(oc get knativeserving.operator.knative.dev knative-serving -n $SERVING_NAMESPACE -o=jsonpath="{.status.version}") == "$serving_version" ]] || return 1
+    else
+      approve_csv "$upgrade_to" "$OLM_UPGRADE_CHANNEL" || return 1
+      timeout 900 '[[ ! ( $(oc get knativeserving.operator.knative.dev knative-serving -n $SERVING_NAMESPACE -o=jsonpath="{.status.version}") != $serving_version && $(oc get knativeserving.operator.knative.dev knative-serving -n $SERVING_NAMESPACE -o=jsonpath="{.status.conditions[?(@.type==\"Ready\")].status}") == True ) ]]' || return 1
+
+      # Assert that the old image references eventually fade away
+      timeout 900 "oc get pod -n $SERVING_NAMESPACE -o yaml | grep image: | uniq | grep $serving_version" || return 1
+    fi
+    end_prober_test ${PROBER_PID}
+  fi
+
+  # Might not work in OpenShift CI but we want it here so that we can consume this script later and re-use
+  if [[ $UPGRADE_CLUSTER == true ]]; then
+    # End the prober test now before we start cluster upgrade, up until now we should have zero failed requests
+    end_prober_test ${PROBER_PID}
```

Comment:
Let's move the `end_prober_test` function definition to this file from `lib.bash` as well.

Reply:
Okay. Will do that with follow up, also with name change to `end_serving_prober_test`, as I hope knative/eventing#2388 will get somewhere, hopelly!

Reply:
On second thought. The eventing upgrade test use the same file as a signal to end probe test, so maybe I'll not move that function.

Reply:
hmm, ok